ipn/ipnlocal/state_test.go

```diff
@@ -0,0 +1,778 @@
+	cc.logf("SetPaused=%v", paused)
+	if paused {
+		cc.called("pause")
+	} else {
+		cc.called("unpause")
+	}
+}
+
+func (cc *mockControl) AuthCantContinue() bool {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+
+	return cc.authBlocked
+}
+
+func (cc *mockControl) SetHostinfo(hi *tailcfg.Hostinfo) {
+	cc.logf("SetHostinfo: %v", *hi)
+	cc.called("SetHostinfo")
+}
+
+func (cc *mockControl) SetNetInfo(ni *tailcfg.NetInfo) {
+	cc.called("SetNetinfo")
+	cc.logf("SetNetInfo: %v", *ni)
+	cc.called("SetNetInfo")
+}
+
+func (cc *mockControl) UpdateEndpoints(localPort uint16, endpoints []tailcfg.Endpoint) {
+	// validate endpoint information here?
+	cc.logf("UpdateEndpoints: lp=%v ep=%v", localPort, endpoints)
+	cc.called("UpdateEndpoints")
+}
+
+// A very precise test of the sequence of function calls generated by
+// ipnlocal.Local into its controlclient instance, and the events it
+// produces upstream into the UI.
+//
+// [apenwarr] Normally I'm not a fan of "mock" style tests, but the precise
+// sequence of this state machine is so important for writing our multiple
+// frontends, that it's worth validating it all in one place.
+//
+// Any changes that affect this test will most likely require carefully
+// re-testing all our GUIs (and the CLI) to make sure we didn't break
+// anything.
+//
+// Note also that this test doesn't have any timers, goroutines, or duplicate
+// detection. It expects messages to be produced in exactly the right order,
+// with no duplicates, without doing network activity (other than through
+// controlclient, which we fake, so there's no network activity there either).
+//
+// TODO: A few messages that depend on magicsock (which actually might have
+// network delays) are just ignored for now, which makes the test
+// predictable, but maybe a bit less thorough. This is more of an overall
+// state machine test than a test of the wgengine+magicsock integration.
+func TestStateMachine(t *testing.T) {
+	assert := assert.New(t)
+
+	logf := t.Logf
+	store := new(ipn.MemoryStore)
+	e, err := wgengine.NewFakeUserspaceEngine(logf, 0)
+	if err != nil {
+		t.Fatalf("NewFakeUserspaceEngine: %v", err)
+	}
+
+	cc := newMockControl()
+	ccGen := func(opts controlclient.Options) (controlclient.Client, error) {
+		cc.mu.Lock()
+		cc.opts = opts
+		cc.logf = opts.Logf
+		cc.authBlocked = true
+		cc.mu.Unlock()
+
+		cc.called("New")
+		return cc, nil
+	}
+	b, err := NewLocalBackendWithClientGen(logf, "logid", store, e, ccGen)
+	if err != nil {
+		t.Fatalf("NewLocalBackend: %v", err)
+	}
+
+	notifies := notifyThrottler{t: t}
+	notifies.expect(0)
+
+	b.SetNotifyCallback(func(n ipn.Notify) {
+		if n.State != nil ||
+			n.Prefs != nil ||
+			n.BrowseToURL != nil ||
+			n.LoginFinished != nil {
+			logf("\n%v\n\n", n)
+			notifies.put(n)
+		} else {
+			logf("\n(ignored) %v\n\n", n)
+		}
+	})
+
+	// Check that it hasn't called us right away.
+	// The state machine should be idle until we call Start().
+	assert.Equal(cc.getCalls(), []string{})
```

Comment:
These are going to get removed and replaced by normal Go test idioms shortly after you submit this. I believe this has been a bone of contention in the past, so I'm not going to fight it here, just stating what will happen :)

Reply:
If someone manages to do that without making the test twice as long, then that's great, I'll learn something.

If someone changes the code and makes it twice as long just to avoid using an assertion library, I'm going to complain, because go's standard test idioms are frustrating, pointless, and extremely error prone, and I'm tired of them. (I cannot express how exhausted I am about error messages not matching the actual thing that happened, or people not bothering to print out the two things they were comparing, in a useful format for comparison...)

I'm not especially attached to *this* assertion library. It already annoys me a little bit (in particular, I really wish they had spelled the args (actual, expected) or just (a, b) instead of (expected, actual), because the comparisons all feel backwards. I would also like it if it printed a message for each assertion, even if it passed.)

Meanwhile, I'm very happy that Go fixed up streaming test output in test -v mode though. At least that's so much better than what we had when I wrote the e2e_test long ago, where you had to trick it into telling you where in the test it got deadlocked.